build-image/Dockerfile

```diff
@@ -0,0 +1,78 @@
+    rust nightly, clippy, rustfmt, miri, rust-src, rustc-dev, grcov, rust-covfix, \
+    llvm-tools-preview, cargo-contract, xargo, binaryen, parallel, codecov, ink, solang" \
+    io.parity.image.revision="${VCS_REF}" \
+    io.parity.image.documentation="https://github.com/paritytech/cargo-contract/blob/${VCS_REF}/\
+    build-image/README.md" \
+    io.parity.image.created="${BUILD_DATE}"
+
+WORKDIR /builds
+
+RUN set -eux; \
+    apt-get -y update && \
+    apt-get install -y --no-install-recommends zlib1g-dev npm wabt && \
+    npm install --ignore-scripts -g yarn && \
+
+    # `binaryen` is needed by `cargo-contract` for optimizing Wasm files.
+    # We fetch the latest release which contains a Linux binary.
+    curl -L $(curl --silent https://api.github.com/repos/WebAssembly/binaryen/releases \
+    | jq -r '.[0].assets | [.[] | .browser_download_url] | map(select(match("x86_64-linux\\.tar\\.gz$"))) | .[0]' \
+    ) | tar -xz -C /usr/local/bin/ --wildcards --strip-components=2 'binaryen-*/bin/wasm-opt' && \
+
+    # Install LLVM for solang
+    curl -L https://github.com/hyperledger/solang-llvm/releases/download/llvm15-1/llvm15.0-linux-x86-64.tar.xz > llvm15.0-linux-x86-64.tar.xz && \
+    tar Jxf llvm15.0-linux-x86-64.tar.xz && \
+    export PATH=$(pwd)/llvm15.0/bin:$PATH && \
+
+    # The stable toolchain is used to build ink! itself through the use of the
+    # `RUSTC_BOOSTRAP=1` environment variable. We also need to install the
+    # `wasm32-unknown-unknown` target since that's the platform that ink! smart contracts
+    # run on.
+
+    # The 1.69 toolchain is temporarily required to build ink! contracts because of
+    # https://github.com/paritytech/cargo-contract/issues/1139 \
+    rustup toolchain install 1.69 && \
+    rustup target add wasm32-unknown-unknown --toolchain 1.69 && \
+    rustup component add rust-src clippy rustfmt --toolchain 1.69 && \
+    rustup default 1.69 && \
+
+    # `cargo-dylint` and `dylint-link` are dependencies needed to run `cargo-contract`.
+    cargo install cargo-dylint dylint-link && \
+
+    # Install the latest stable release of `cargo-contract`
+    cargo install cargo-contract && \
+
+    # Install the latest `solang`
+    cargo install solang && \
+
+    # Versions
+    rustup show && \
+    cargo --version && \
+    cargo-contract --version && \
+    wasm-opt --version && \
+
+    # Clean up and remove compilation artifacts that a cargo install creates (>250M).
+    rm -rf "${CARGO_HOME}/registry" "${CARGO_HOME}/git" /root/.cache/sccache && \
+
+    # apt clean up
+    apt-get remove -y gnupg && \
+    apt-get autoremove -y && \
+    apt-get clean && \
+    rm -rf /var/lib/apt/lists/*
+
+# A contract workdir which should be bind mounted to the container. See README for detailed documentation.
+WORKDIR /contract
+
+# A default entry point for manual execution
+ENTRYPOINT [ "cargo", "contract", "build", "--release" ]
```

Comment:
I've tried building this Dockerfile on my machine - Mac Mini M2 Pro - and this are my findings:
1. The build took 7732 seconds to finish (that's over 2 hours 😬 ):
```
[+] Building 455.9s (5/7)
[+] Building 456.5s (5/7)
[+] Building 588.1s (5/7)
[+] Building 3495.5s (5/7)
[+] Building 3495.6s (5/7)
[+] Building 3495.8s (5/7)
[+] Building 3495.9s (5/7)
[+] Building 7732.4s (8/8) FINISHED
```
2. The resulting image is over 5GB big:
```
REPOSITORY                            TAG       IMAGE ID       CREATED             SIZE
<none>                                <none>    46ca20145887   21 minutes ago      5.44GB
```
3. They way it's built is not arch-independent (or OS) and when building on Mac I was getting lots of warnings about the allocator, one for every dependency built:
```
=> => #    Compiling is-terminal v0.4.7
 => => # <jemalloc>: MADV_DONTNEED does not work (memset will be used instead)
 => => # <jemalloc>: (This is the expected behaviour if you are running under QEMU)
```
4. At the peak, it used almost 23GB of RAM. At the beginning my OS was killing the process when it hit the initial RAM ceiling (16GB).

Consider taking a look at our [Dockerfile](https://github.com/Cardinal-Cryptography/docker-ink-dev/blob/main/Dockerfile) and the accompanying [Makefile](https://github.com/Cardinal-Cryptography/docker-ink-dev/blob/main/Makefile).

For comparison:
1. Build time was 797s:
```
➜  docker-ink-dev git:(main) ✗ make build-image
docker buildx build --pull --platform linux/arm64/v8  -t cardinal-cryptography/ink-dev-arm64:1.1.0 --load . \
	&& docker tag cardinal-cryptography/ink-dev-arm64:1.1.0 cardinal-cryptography/ink-dev-arm64:latest
[+] Building 534.8s (11/16)
[+] Building 545.7s (11/16)
[+] Building 797.5s (17/17) FINISHED
```
2. Image size is slightly over 1GB (this image contains (ink-wrapper)[ink-wrapper/aleph_client.rs at main · Cardinal-Cryptography/ink-wrapper], w/o it's 800MB):
```
gorskimateusz@colima:/Users/gorskimateusz$ docker images
REPOSITORY                            TAG       IMAGE ID       CREATED          SIZE
cardinal-cryptography/ink-dev-arm64   1.1.0     0f3f4874078e   19 seconds ago   1.13GB
cardinal-cryptography/ink-dev-arm64   latest    0f3f4874078e   19 seconds ago   1.13GB
```
3. Makefile instructions detect the underlying architecture and use correct flags when building so no warnings are logged and the process is much faster.
4. At peak, the process consumed ~1.3GB of RAM.

Reply:
Yes, this is because you are using ARM64 base when building an image on ARM host. To guarantee the _absolute_ determinism in builds. We need to use a completely identical architecture for the builds. That's why we are sticking with AMD64 base image. In fact, the user won't be required to build the image. We are gonna publish on docker hub.  Deploying and running the container from the pre-built image takes ~1-2m depending on the complexity of the contract.

Additionally, the image contains solang and ask! dependencies as we also need to provide a deterministic environment for these languages.

I will take a look at the dockerfile again and see how we can see reduce the size.

Reply:
I can imagine that there will have to exist a docker image for rust toolchain + cargo contract combination as each can change the outcome. Will you be publishing docker images for every one?

Reply:
we aim to use latest rust toolchain and stable releases of ink! and cargo-contract. The contract's metadata file will contain the tag of the image used to compile it, so that block explorers can pull the image and verify the contract with it.

Whenever there are updates, we are just gonna publish a new tag.

Reply:
But contracts can choose to use their own toolchain and override yours with rust-toolchain file. Are you going to ignore that?

Reply:
We don't ignore that. We persist the toolchain and custom compilation flags used in the metadata of the contract.

Reply:
I'm confirming with ask! and solang devs, if they require deterministic builds. If not, then we are likely to adopt slim, platform-agnostic rust image that your dockerfile uses.

Reply:
> We don't ignore that. We persist the toolchain and custom compilation flags used in the metadata of the contract.

You wrote earlier that you aim to use latest toolchain, stable release of ink! and stable release of cargo-contract. That's a triple. If you say you will publish a new docker image every time one of the three release new stable version then that's OK - that's basically a _yes_ to my previous question:
> Will you be publishing docker images for every one?